librz/analysis/dwarf_process.c

```diff
@@ -96,93 +70,110 @@ static inline char *create_type_name_from_offset(ut64 offset) {
-		const char *s = rz_bin_dwarf_attr_value_get_string_content(&die->attr_values[name_attr_idx]);
+	if (attr) {
+		const char *s = rz_bin_dwarf_attr_value_get_string_content(attr);
 		name = RZ_STR_DUP(s);
 	}
 	return name ? name : create_type_name_from_offset(die->offset);
 }
 
+static RzPVector /*<RzBinDwarfDie *>*/ *die_children(RzBinDwarfDie *die, RzBinDwarf *dw) {
+	RzPVector /*<RzBinDwarfDie *>*/ *vec = rz_pvector_new(NULL);
+	if (!vec) {
+		return NULL;
+	}
+	RzBinDwarfCompUnit *unit = ht_up_find(dw->info->unit_tbl, die->unit_offset, NULL);
+	if (!unit) {
+		goto err;
+	}
+
+	for (size_t i = die->index + 1; i < rz_vector_len(&unit->dies); ++i) {
+		RzBinDwarfDie *child_die = rz_vector_index_ptr(&unit->dies, i);
+		if (child_die->depth >= die->depth + 1) {
+			rz_pvector_push(vec, child_die);
+		} else if (child_die->depth == die->depth) {
+			break;
+		}
+	}
+
+	return vec;
+err:
+	rz_pvector_free(vec);
+	return NULL;
+}
+
 /**
  * \brief Get the DIE size in bits
  *
  * \param die
  * \return ut64 size in bits or 0 if not found
  */
-static ut64 get_die_size(const RzBinDwarfDie *die) {
-	ut64 size = 0;
-	st32 byte_size_idx = find_attr_idx(die, DW_AT_byte_size);
-
-	if (byte_size_idx != -1) {
-		size = die->attr_values[byte_size_idx].uconstant * CHAR_BIT;
-	} else {
-		st32 bit_size_idx = find_attr_idx(die, DW_AT_bit_size);
+static ut64 die_byte_size(const RzBinDwarfDie *die) {
+	RzBinDwarfAttr *attr = rz_bin_dwarf_die_get_attr(die, DW_AT_byte_size);
+	if (attr) {
+		return attr->uconstant * CHAR_BIT;
+	}
 
-		if (bit_size_idx != -1) {
-			size = die->attr_values[bit_size_idx].uconstant;
-		}
+	attr = rz_bin_dwarf_die_get_attr(die, DW_AT_bit_size);
+	if (attr) {
+		return attr->uconstant;
 	}
-	return size;
+
+	return 0;
 }
 
 /**
  * \brief Parse and return the count of an array or 0 if not found/not defined
  */
-static ut64 parse_array_count(Context *ctx, ut64 idx) {
-	const RzBinDwarfDie *die = &ctx->all_dies[idx++];
-
-	if (die->has_children) {
-		int child_depth = 1;
-		size_t j;
-		for (j = idx; child_depth > 0 && j < ctx->count; j++) {
-			const RzBinDwarfDie *child_die = &ctx->all_dies[j];
-			// right now we skip non direct descendats of the structure
-			// can be also DW_TAG_suprogram for class methods or tag for templates
-			if (child_depth == 1 && child_die->tag == DW_TAG_subrange_type) {
-				size_t i;
-				for (i = 0; i < child_die->count; i++) {
-					const RzBinDwarfAttrValue *value = &child_die->attr_values[i];
-					switch (value->attr_name) {
-					case DW_AT_upper_bound:
-					case DW_AT_count:
-						return value->uconstant + 1;
-						break;
-					default:
-						break;
-					}
+static ut64 parse_array_count(Context *ctx, RzBinDwarfDie *die) {
+	if (!die->has_children) {
+		return 0;
+	}
+	RzPVector *children = die_children(die, ctx->dw);
+	if (!children) {
+		return 0;
+	}
+
+	void **it;
+	rz_pvector_foreach (children, it) {
+		RzBinDwarfDie *child_die = *it;
+		if (child_die->tag == DW_TAG_subrange_type) {
```

Comment:
If you reverse the condition here, and use `continue`, it will reduce the nesting by one level.